## The X-files problem

In [404]:
import pandas as pd
import numpy as np

In [510]:
df = pd.read_csv('ufo_sightings/UFO_sightings.csv')
df

/Users/demetrio/opt/anaconda3/lib/python3.7/site-packages/IPython/core/interactiveshell.py:3166: DtypeWarning: Columns (5,9) have mixed types.Specify dtype option on import or set low_memory=False.
  interactivity=interactivity, compiler=compiler, result=result)


datetime                  city state country     shape  \
0      10/10/1949 20:30            san marcos    tx      us  cylinder   
1      10/10/1949 21:00          lackland afb    tx     NaN     light   
2      10/10/1955 17:00  chester (uk/england)   NaN      gb    circle   
3      10/10/1956 21:00                  edna    tx      us    circle   
4      10/10/1960 20:00               kaneohe    hi      us     light   
...                 ...                   ...   ...     ...       ...   
80327    9/9/2013 21:15             nashville    tn      us     light   
80328    9/9/2013 22:00                 boise    id      us    circle   
80329    9/9/2013 22:00                  napa    ca      us     other   
80330    9/9/2013 22:20                vienna    va      us    circle   
80331    9/9/2013 23:00                edmond    ok      us     cigar   

      duration (seconds) duration (hours/min)  \
0                   2700           45 minutes   
1                   7200              1-2 hrs   
2                     20           20 seconds   
3                     20             1/2 hour   
4                    900           15 minutes   
...                  ...                  ...   
80327              600.0           10 minutes   
80328             1200.0           20 minutes   
80329             1200.0                 hour   
80330                5.0            5 seconds   
80331             1020.0           17 minutes   

                                                comments date posted  \
0      This event took place in early fall around 194...   4/27/2004   
1      1949 Lackland AFB&#44 TX.  Lights racing acros...  12/16/2005   
2      Green/Orange circular disc over Chester&#44 En...   1/21/2008   
3      My older brother and twin sister were leaving ...   1/17/2004   
4      AS a Marine 1st Lt. flying an FJ4B fighter/att...   1/22/2004   
...                                                  ...         ...   
80327  Round from the distance/slowly changing colors...   9/30/2013   
80328  Boise&#44 ID&#44 spherical&#44 20 min&#44 10 r...   9/30/2013   
80329                                       Napa UFO&#44   9/30/2013   
80330  Saw a five gold lit cicular craft moving fastl...   9/30/2013   
80331  2 witnesses 2  miles apart&#44 Red &amp; White...   9/30/2013   

         latitude  longitude   
0      29.8830556  -97.941111  
1        29.38421  -98.581082  
2            53.2   -2.916667  
3      28.9783333  -96.645833  
4      21.4180556 -157.803611  
...           ...         ...  
80327   36.165833  -86.784444  
80328   43.613611 -116.202500  
80329   38.297222 -122.284444  
80330   38.901111  -77.265556  
80331   35.652778  -97.477778  

[80332 rows x 11 columns]

In [511]:
data_info = pd.read_excel('./ufo_sightings/data_dictionary.xlsx')
data_info

datetime                          date and time of event  \
0                  city                                Name of the city   
1                 state                             State code of event   
2               country                           Country code of event   
3                 shape                                Shape of the UFO   
4    duration (seconds)            Durantion of the sighting in seconds   
5  duration (hours/min)  Durantion of the sighting in hours and minutes   
6              comments                        description of the event   
7           date posted                Date when the event was reported   
8              latitude                            Latitude of the city   
9             longitude                           Longitude of the city   

    String  
0   String  
1   String  
2   String  
3   String  
4  Numeric  
5   String  
6   String  
7     Date  
8  Numeric  
9  Numeric

In [512]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 80332 entries, 0 to 80331
Data columns (total 11 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   datetime              80332 non-null  object 
 1   city                  80332 non-null  object 
 2   state                 74535 non-null  object 
 3   country               70662 non-null  object 
 4   shape                 78400 non-null  object 
 5   duration (seconds)    80332 non-null  object 
 6   duration (hours/min)  80332 non-null  object 
 7   comments              80317 non-null  object 
 8   date posted           80332 non-null  object 
 9   latitude              80332 non-null  object 
 10  longitude             80332 non-null  float64
dtypes: float64(1), object(10)
memory usage: 6.7+ MB


## Cleaning Data

Empezamos estudiando los datos que tenemos

In [513]:
df.duplicated().value_counts()
# No hay valores duplicados

False    80332
dtype: int64

In [514]:
df.isnull().any()

datetime                False
city                    False
state                    True
country                  True
shape                    True
duration (seconds)      False
duration (hours/min)    False
comments                 True
date posted             False
latitude                False
longitude               False
dtype: bool

In [515]:
df.isnull().sum()

datetime                   0
city                       0
state                   5797
country                 9670
shape                   1932
duration (seconds)         0
duration (hours/min)       0
comments                  15
date posted                0
latitude                   0
longitude                  0
dtype: int64

Entonces, tenemos valores nulos en las variables de: state(5797), country(9670), shape(1932) y comments(15).
Ahora, será necesario identificar los renglones donde estos se encuentren.

In [498]:
nan_rows = df[df.isnull().any(1)]
nan_rows

datetime                          city state country     shape  \
1      10/10/1949 21:00                  lackland afb    tx     NaN     light   
2      10/10/1955 17:00          chester (uk/england)   NaN      gb    circle   
6      10/10/1965 21:00            penarth (uk/wales)   NaN      gb    circle   
18     10/10/1973 23:00                   bermuda nas   NaN     NaN     light   
20     10/10/1974 21:30            cardiff (uk/wales)   NaN      gb      disk   
...                 ...                           ...   ...     ...       ...   
80254    9/9/2009 21:15  nottinghamshire (uk/england)   NaN      gb  fireball   
80255    9/9/2009 21:38       kaiserlautern (germany)   NaN      de     light   
80319    9/9/2013 20:15                       clifton    nj     NaN     other   
80322    9/9/2013 21:00          aleksandrow (poland)   NaN     NaN     light   
80324    9/9/2013 21:00         hamstead (hollyridge)    nc     NaN     light   

      duration (seconds) duration (hours/min)  \
1                   7200              1-2 hrs   
2                     20           20 seconds   
6                    180         about 3 mins   
18                    20              20 sec.   
20                  1200           20 minutes   
...                  ...                  ...   
80254              600.0              10 mins   
80255               40.0     about 40 seconds   
80319             3600.0                ~1hr+   
80322               15.0           15 seconds   
80324              120.0            2 minutes   

                                                comments date posted  \
1      1949 Lackland AFB&#44 TX.  Lights racing acros...  12/16/2005   
2      Green/Orange circular disc over Chester&#44 En...   1/21/2008   
6      penarth uk  circle  3mins  stayed 30ft above m...   2/14/2006   
18     saw fast moving blip on the radar scope thin w...   1/11/2002   
20     back in 1974 I was 19 at the time and  lived i...    2/1/2007   
...                                                  ...         ...   
80254  resembled orange flame imagine a transparent h...  12/12/2009   
80255  2 white lights over Kaiserslautern&#44 ramstei...  12/12/2009   
80319              Luminous line seen in New Jersey sky.   9/30/2013   
80322  Two points of light following one another in a...   9/30/2013   
80324  8 to ten lights bright orange in color large t...   9/30/2013   

        latitude  longitude   
1       29.38421  -98.581082  
2           53.2   -2.916667  
6      51.434722   -3.180000  
18     32.364167  -64.678611  
20          51.5   -3.200000  
...          ...         ...  
80254  53.166667   -1.000000  
80255      49.45    7.750000  
80319  40.858433  -74.163755  
80322  50.465843   22.891814  
80324  34.367594  -77.710548  

[13816 rows x 11 columns]

Con eso, extraemos las filas con valores NaN

In [499]:
df.columns[df.isnull().any()]

Index(['state', 'country', 'shape', 'comments'], dtype='object')

Después de ver los datos y entender mejor el problema. Considero que los resultados que debemos entregar responde a las siguientes preguntas:

¿Cuáles son las ciudades, meses, y horas con mayor avistamiento? Así como sus coordenadas. Las respuestas a éstas preguntas son las que llevará a la gente al lugar indicado para obserbar OVNI's.

Analizando las columnas las variables que necesitamos estudiar son: 'datatime', 'city', 'latitude' y 'longitude' principalmente, sin embargo estos no contienen valores nulos. Entonces, podemos hacer el análisis con los datos completos.

Ahora, es necesario limpiar los datos y dejarlos con el tipo de dato más efectivo para el análisis.

### Cambiar la columna 'datatime' a tipo de dato datatime.

In [500]:
df['datetime'].sample(50)

59946     7/26/2007 22:30
7355      10/9/2011 19:42
67909     8/20/1985 19:30
54650     7/10/1997 16:00
16782     1/20/2009 19:00
49895     6/18/2003 17:30
62426      7/4/2012 20:40
36788     4/15/1968 02:00
24948      1/6/2012 23:00
66825     8/15/2012 19:35
37508     4/17/2014 14:00
22414     12/4/2012 20:00
43656      5/2/1978 21:00
10471    11/18/2006 22:50
41008      4/9/2004 22:30
36282     4/12/2002 22:30
30493     3/11/2008 22:00
10430    11/18/2003 18:35
42159     5/14/1995 20:37
58643     7/21/2009 21:30
75043     9/16/1984 15:00
57722     7/18/1970 23:00
37167     4/16/2004 13:20
57508     7/17/1998 22:51
42400     5/15/1993 17:30
25506     2/10/1979 23:00
45518      5/4/2007 23:30
53248     6/30/2012 23:00
52888     6/30/1993 21:00
37375     4/17/2002 20:55
20835    12/27/2002 23:30
43590     5/21/2007 24:00
58805      7/2/2005 03:30
28729     2/26/2013 17:30
31099     3/13/2001 18:20
4166     10/23/2012 20:00
58550     7/21/2002 22:52
23241     1/27/2007 00:00
59451     7/

Haciendo la conversión.

In [516]:
df['datetime'] = pd.to_datetime(df['datetime'])
df['datetime']

ParserError: hour must be in 0..23: 10/11/2006 24:00

M/D/Y H/M

El error nos indica que el intervalo para escribir la hora es de 0 a 23. Entonces, algún valor aparece como 24:00, en vez de 00:00.

Exploremos la columna, para identificar donde se encuentra.

In [517]:
pd.to_datetime(df.iloc[1:388,0])

1     1949-10-10 21:00:00
2     1955-10-10 17:00:00
3     1956-10-10 21:00:00
4     1960-10-10 20:00:00
5     1961-10-10 19:00:00
              ...        
383   2006-10-11 21:00:00
384   2006-10-11 21:20:00
385   2006-10-11 21:32:00
386   2006-10-11 22:00:00
387   2006-10-11 23:00:00
Name: datetime, Length: 387, dtype: datetime64[ns]

In [518]:
df.iloc[380:390,0]

380    10/11/2006 19:00
381    10/11/2006 19:30
382    10/11/2006 20:45
383    10/11/2006 21:00
384    10/11/2006 21:20
385    10/11/2006 21:32
386    10/11/2006 22:00
387    10/11/2006 23:00
388    10/11/2006 24:00
389    10/11/2007 04:40
Name: datetime, dtype: object

 Cambiamos el valor.

In [519]:
df.iloc[388,0] = '10/11/2006 00:00'
df.iloc[388,0]

'10/11/2006 00:00'

Ahora, ya podría convetir toda la columna datetime a tipo de dato datetime

In [520]:
df['datetime'] = pd.to_datetime(df['datetime'])
df['datetime']

ParserError: hour must be in 0..23: 10/1/2001 24:00

Al parecer, hay más valores con 24:00.
Debemos econtrarlos todos.

In [521]:
pd.to_datetime(df.iloc[1:693,0])

1     1949-10-10 21:00:00
2     1955-10-10 17:00:00
3     1956-10-10 21:00:00
4     1960-10-10 20:00:00
5     1961-10-10 19:00:00
              ...        
688   2001-10-01 20:35:00
689   2001-10-01 21:00:00
690   2001-10-01 21:00:00
691   2001-10-01 22:00:00
692   2001-10-01 22:30:00
Name: datetime, Length: 692, dtype: datetime64[ns]

In [522]:
df.iloc[690:700,0]

690    10/1/2001 21:00
691    10/1/2001 22:00
692    10/1/2001 22:30
693    10/1/2001 24:00
694    10/1/2002 00:30
695    10/1/2002 00:30
696    10/1/2002 03:00
697    10/1/2002 05:30
698    10/1/2002 05:30
699    10/1/2002 05:30
Name: datetime, dtype: object

In [523]:
df.iloc[693,0] = '10/1/2001 00:00'

In [524]:
df['datetime'] = pd.to_datetime(df['datetime'])
df['datetime']

ParserError: hour must be in 0..23: 10/1/2012 24:00

Al parecer hay muchos más valores, entonces, haré el cambio de manera general.

Encotrar todos los valores con hora dado por '24:' y cambiarlo por '00:'

In [525]:
df['datetime'] = df['datetime'].str.replace('24:','00:')

Cambiamos los '24:' a '00:' en toda la columna.

In [527]:
df['datetime'] = pd.to_datetime(df['datetime'])
df['datetime']

0       1949-10-10 20:30:00
1       1949-10-10 21:00:00
2       1955-10-10 17:00:00
3       1956-10-10 21:00:00
4       1960-10-10 20:00:00
                ...        
80327   2013-09-09 21:15:00
80328   2013-09-09 22:00:00
80329   2013-09-09 22:00:00
80330   2013-09-09 22:20:00
80331   2013-09-09 23:00:00
Name: datetime, Length: 80332, dtype: datetime64[ns]

Con esto, hemos convertido a toda la columna en tipo de dato datetime.

In [528]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 80332 entries, 0 to 80331
Data columns (total 11 columns):
 #   Column                Non-Null Count  Dtype         
---  ------                --------------  -----         
 0   datetime              80332 non-null  datetime64[ns]
 1   city                  80332 non-null  object        
 2   state                 74535 non-null  object        
 3   country               70662 non-null  object        
 4   shape                 78400 non-null  object        
 5   duration (seconds)    80332 non-null  object        
 6   duration (hours/min)  80332 non-null  object        
 7   comments              80317 non-null  object        
 8   date posted           80332 non-null  object        
 9   latitude              80332 non-null  object        
 10  longitude             80332 non-null  float64       
dtypes: datetime64[ns](1), float64(1), object(9)
memory usage: 6.7+ MB


Ahora, analizemos la columna de 'city'

In [668]:
df['city'].unique()

array(['san marcos', 'lackland afb', 'chester (uk/england)', ...,
       'calmar (canada)', 'aleksandrow (poland)', 'hamstead (hollyridge)'],
      dtype=object)

In [670]:
df_new['city'].value_counts().head(60)

seattle                     525
phoenix                     454
portland                    374
las vegas                   368
los angeles                 353
san diego                   338
houston                     297
chicago                     265
tucson                      241
miami                       239
orlando                     220
austin                      218
springfield                 214
albuquerque                 213
sacramento                  201
columbus                    200
san jose                    189
london (uk/england)         189
san francisco               187
denver                      185
san antonio                 175
jacksonville                172
toronto (canada)            172
myrtle beach                166
dallas                      160
salem                       154
charlotte                   148
new york city (brooklyn)    147
columbia                    147
arlington                   142
mesa                        139
louisvil

In [636]:
#df['latitude'] = pd.to_numeric(df['latitude'])
df.iloc[43780:43790, 9:10]

latitude
43780  37.7975000
43781  41.4047222
43782  33q.200088
43783  32.4366667
43784  34.7463889
43785   59.119149
43786  39.0483333
43787  32.2216667
43788  40.7544444
43789  27.0994444

Hay un error en la posición 43782. Dice '33q200088'. La latitud, son valores que van de 0 a 90, entonces esperaría que la letra 'q', sólo sea un error extra y no un número faltante. Así que, sólo lo eliminaremos.

In [640]:
df['latitude'][43782] = '33.200088'

/Users/demetrio/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  """Entry point for launching an IPython kernel.


In [643]:
df['latitude'] = pd.to_numeric(df['latitude'])
df['latitude']

0        29.883056
1        29.384210
2        53.200000
3        28.978333
4        21.418056
           ...    
80327    36.165833
80328    43.613611
80329    38.297222
80330    38.901111
80331    35.652778
Name: latitude, Length: 80332, dtype: float64

Hay un espacio al final del nombre de columna longitude, así que, remplazemoslo por:

In [652]:
df['longitude'] = df['longitude ']
df['longitude']

0        -97.941111
1        -98.581082
2         -2.916667
3        -96.645833
4       -157.803611
            ...    
80327    -86.784444
80328   -116.202500
80329   -122.284444
80330    -77.265556
80331    -97.477778
Name: longitude, Length: 80332, dtype: float64

Así, terminaríamos con una tabla para las 4 variables más importantes para responder a nuestras preguntas.

In [671]:
df_new = df[['datetime', 'city', 'latitude', 'longitude']]
df_new

datetime                  city   latitude   longitude
0     1949-10-10 20:30:00            san marcos  29.883056  -97.941111
1     1949-10-10 21:00:00          lackland afb  29.384210  -98.581082
2     1955-10-10 17:00:00  chester (uk/england)  53.200000   -2.916667
3     1956-10-10 21:00:00                  edna  28.978333  -96.645833
4     1960-10-10 20:00:00               kaneohe  21.418056 -157.803611
...                   ...                   ...        ...         ...
80327 2013-09-09 21:15:00             nashville  36.165833  -86.784444
80328 2013-09-09 22:00:00                 boise  43.613611 -116.202500
80329 2013-09-09 22:00:00                  napa  38.297222 -122.284444
80330 2013-09-09 22:20:00                vienna  38.901111  -77.265556
80331 2013-09-09 23:00:00                edmond  35.652778  -97.477778

[80332 rows x 4 columns]

In [681]:
df_new.groupby('city')['latitude']